lib/Dialect/AIE/Transforms/AIEGenerateColumnControlOverlay.cpp

```diff
@@ -0,0 +1,255 @@
+        SmallVector<AIE::TileOp> tilesOnCol;
+        for (auto &[tId, tOp] : tiles) {
+          if (tId.col != col)
+            continue;
+          if (targetModel.isCoreTile(tId.col, tId.row) ||
+              targetModel.isMemTile(tId.col, tId.row))
+            tilesOnCol.push_back(tOp);
+        }
+
+        int unusedPacketIdFrom = designUnusedPacketIdFrom;
+        int numShimChans = targetModel.getNumSourceShimMuxConnections(
+            shimTile.getCol(), shimTile.getRow(), AIE::WireBundle::DMA);
+        SmallVector<int> availableShimChans = getAvailableShimChans(
+            device, numShimChans, shimTile, AIE::WireBundle::DMA, true);
+        // Create packet flows per col
+        generatePacketFlowsForControl(builder, device, targetModel, shimTile,
+                                      AIE::WireBundle::DMA, availableShimChans,
+                                      tilesOnCol, AIE::WireBundle::Ctrl, 0,
+                                      unusedPacketIdFrom, true);
+      }
+    }
+  }
+
+  AIE::PacketFlowOp
+  createPacketFlowOp(OpBuilder &builder, int &flowID, Value source,
+                     xilinx::AIE::WireBundle sourceBundle,
+                     uint32_t sourceChannel, Value dest,
+                     xilinx::AIE::WireBundle destBundle, uint32_t destChannel,
+                     mlir::BoolAttr keep_pkt_header = nullptr) {
+    AIE::PacketFlowOp pktFlow = builder.create<AIE::PacketFlowOp>(
+        builder.getUnknownLoc(), flowID++, keep_pkt_header);
+    Region &r_pktFlow = pktFlow.getPorts();
+    Block *b_pktFlow = builder.createBlock(&r_pktFlow);
+    builder.setInsertionPointToStart(b_pktFlow);
+    builder.create<AIE::PacketSourceOp>(builder.getUnknownLoc(), source,
+                                        sourceBundle, sourceChannel);
+    builder.create<AIE::PacketDestOp>(builder.getUnknownLoc(), dest, destBundle,
+                                      destChannel);
+    builder.create<AIE::EndOp>(builder.getUnknownLoc());
+    return pktFlow;
+  }
+
+  // Create packet flows per col which moves control packets to and from shim
+  // dma
+  void generatePacketFlowsForControl(OpBuilder builder, DeviceOp device,
+                                     const AIETargetModel &targetModel,
+                                     TileOp shimTile, WireBundle shimWireBundle,
+                                     SmallVector<int> availableShimChans,
+                                     SmallVector<AIE::TileOp> ctrlTiles,
+                                     WireBundle ctrlWireBundle,
+                                     int coreOrMemChanId,
+                                     int unusedPacketIdFrom, bool isShimMM2S) {
+    // Create packet flows
+    SmallVector<int> thresholdsToNextShimChannel;
+    for (int i = 1; i < (int)availableShimChans.size() + 1; i++)
+      thresholdsToNextShimChannel.push_back(ctrlTiles.size() /
+                                            availableShimChans.size() * i);
```

Comment:
Can you add comments explaining what `thresholdsToNextShimChannel` is doing?

Reply:
Sure, comment added.